### DISCUSSION TITLE
Normalisation in FFN of Primus transformer block

### DISCUSSION THREAD

**Discussion by xdoni4** [MIC-DKFZ/ssl3d#7] [Q&A]
Dear SSL 3D challenge organisers,

`Primus` arhitecture utilizes timm `Eva` transformer implementation under the hood, providing some kwargs that can be passed to Eva initialisation. Eva implemetation has `scale_mlp` parameter that controls whether normalisation in `EvaBlock` FFN is used or not. By default it is set to `True`, and as `Primus` does not modify it during `Eva` initialisation, FFN layers in `Primus` blocks perform normalisation.

Example: timm SwiGLU, where normalisation is not `nn.Identity` when `scale_mlp=True`

```python
class SwiGLU(nn.Module):
    """ SwiGLU
    NOTE: GluMLP above can implement SwiGLU, but this impl has split fc1 and
    better matches some other common impl which makes mapping checkpoints simpler.
    """
    def __init__(
            self,
            in_features,
            hidden_features=None,
            out_features=None,
            act_layer=nn.SiLU,
            norm_layer=None,
            bias=True,
            drop=0.,
    ):
        super().__init__()
        out_features = out_features or in_features
        hidden_features = hidden_features or in_features
        bias = to_2tuple(bias)
        drop_probs = to_2tuple(drop)

        self.fc1_g = nn.Linear(in_features, hidden_features, bias=bias[0])
        self.fc1_x = nn.Linear(in_features, hidden_features, bias=bias[0])
        self.act = act_layer()
        self.drop1 = nn.Dropout(drop_probs[0])
        self.norm = norm_layer(hidden_features) if norm_layer is not None else nn.Identity()
        self.fc2 = nn.Linear(hidden_features, out_features, bias=bias[1])
        self.drop2 = nn.Dropout(drop_probs[1])

    def init_weights(self):
        # override init of fc1 w/ gate portion set to weight near zero, bias=1
        if self.fc1_g.bias is not None:
            nn.init.ones_(self.fc1_g.bias)
        nn.init.normal_(self.fc1_g.weight, std=1e-6)

    def forward(self, x):
        x_gate = self.fc1_g(x)
        x = self.fc1_x(x)
        x = self.act(x_gate) * x
        x = self.drop1(x)
        x = self.norm(x)
        x = self.fc2(x)
        x = self.drop2(x)
        return x
```

Many profound SSL image  transformer architectures (e.g. SimMIM, iBOT, DINO, MAE) deliberately do not use normalisation in FFN. Personally, I encountered performance drops on downstream objectives when switching to FFN with normalisation, which suggests that adding normalisation in FFN may harm pretraining. Unfortunately, adaptation from architecture with no FFN normalisation to architecture with FFN normalisation is not possible: those layers that used to receive unnormalized features will not act adequately.

Considering the above, is it possible to allow using implementations with no FFN normalisation? It can be done without changing the overall Primus architecture, just adding `scale_mlp` as a parameter:
```python
class Primus(AbstractDynamicNetworkArchitectures):

    def __init__(
        self,
        input_channels: int,
        embed_dim: int,
        patch_embed_size: Tuple[int, ...],
        num_classes: int,
        eva_depth: int = 24,
        eva_numheads: int = 16,
        input_shape: Tuple[int, ...] = None,
        decoder_norm=LayerNormNd,
        decoder_act=nn.GELU,
        num_register_tokens: int = 0,
        use_rot_pos_emb: bool = True,
        use_abs_pos_embed: bool = True,
        mlp_ratio=4 * 2 / 3,
        drop_path_rate=0,  # drops computations (multihead attention, mlp), Implementation of scaling might be useless here because this is not batch normed
        patch_drop_rate: float = 0.0,  # drops input patches, may be used for MAE style pretraining
        proj_drop_rate: float = 0.0,  # drops out things related to the projection. That is in the MLP and at the end of EVA attention
        attn_drop_rate: float = 0.0,  # drops attention, meaning connections between patches may bebroken up at random
        rope_impl=RotaryEmbeddingCat,
        rope_kwargs=None,
        init_values=None,
        scale_attn_inner=False,
        scale_mlp=True # add this, True preserves current Primus, False allows for no norm in FFN
    ):
```
```python
# we need to compute the ref_feat_shape for eva
        self.eva = Eva(
            embed_dim=embed_dim,
            depth=eva_depth,
            num_heads=eva_numheads,
            ref_feat_shape=tuple([i // ds for i, ds in zip(input_shape, patch_embed_size)]),
            num_reg_tokens=num_register_tokens,
            use_rot_pos_emb=use_rot_pos_emb,
            use_abs_pos_emb=use_abs_pos_embed,
            mlp_ratio=mlp_ratio,
            drop_path_rate=drop_path_rate,
            patch_drop_rate=patch_drop_rate,
            proj_drop_rate=proj_drop_rate,
            attn_drop_rate=attn_drop_rate,
            rope_impl=rope_impl,
            rope_kwargs=rope_kwargs,
            init_values=init_values,
            scale_attn_inner=scale_attn_inner,
            scale_mlp=scale_mlp # and add this
        )
```

**Comment by TaWald** [comment]
Hey @xdoni4 ,
While removing the normalization may help in improving your performance, we deliberately chose to keep the architecture fully frozen. 
This is largely intended to have participants of the SSL3D challenge focus _purely on pretraining method advances_ as opposed to increasing the search space to other things like modifying the architecture. 

Hence, we won't allow changing this parameter as it is out-of-scope of the challenge and as it would be unfair to other participants who adhered to the initial architectural choices. 

Hope this doesn't impact your developments too much. 
Best, Tassilo
